compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -34,53 +37,180 @@ pub fn expand_deriving_default(
+        }
+        [first, rest @ ..] => {
+            cx.struct_span_err(trait_span, "multiple declared defaults")
+                .span_label(first.span, "first default variant")
+                .span_labels(rest.iter().map(|variant| variant.span), "additional default variant")
+                .note("only one variant can be default")
+                .emit();
+
+            return Err(());
+        }
+    };
+
+    if !matches!(variant.data, VariantData::Unit(..)) {
+        cx.struct_span_err(variant.ident.span, "`#[default]` may only be used on unit variants")
+            .help("consider a manual implementation of `Default`")
             .emit();
-            // let compilation continue
-            DummyResult::raw_expr(trait_span, true)
+
+        return Err(());
+    }
+
+    if let Some(non_exhaustive_attr) =
+        variant.attrs.iter().find(|attr| attr.has_name(sym::non_exhaustive))
+    {
+        cx.struct_span_err(variant.ident.span, "default variant must be exhaustive")
+            .span_label(non_exhaustive_attr.span, "variant declared `#[non_exhaustive]` here")
+            .help("consider a manual implementation of `Default`")
+            .emit();
+
+        return Err(());
+    }
+
+    Ok(variant)
+}
+
+fn validate_default_attribute(
+    cx: &mut ExtCtxt<'_>,
+    default_variant: &rustc_ast::Variant,
+) -> Result<(), ()> {
+    let attrs: Vec<_> =
+        default_variant.attrs.iter().filter(|attr| attr.has_name(kw::Default)).collect();
+
+    let attr = match attrs.as_slice() {
+        [attr] => attr,
+        [] => cx.bug(
+            "this method must only be called with a variant that has a `#[default]` attribute",
+        ),
+        [first, rest @ ..] => {
+            // FIXME Do we want to perform this check? It doesn't exist for `#[inline]`,
+            // `#[non_exhaustive]`, and presumably others.
+
+            let suggestion_text =
+                if rest.len() == 1 { "try removing this" } else { "try removing these" };
+
+            cx.struct_span_err(default_variant.ident.span, "multiple `#[default]` attributes")
+                .note("only one `#[default]` attribute is needed")
+                .span_label(first.span, "`#[default]` used here")
+                .span_label(rest[0].span, "`#[default]` used again here")
+                .span_help(rest.iter().map(|attr| attr.span).collect::<Vec<_>>(), suggestion_text)
+                // This would otherwise display the empty replacement, hence the otherwise
+                // repetitive `.span_help` call above.
+                .tool_only_multipart_suggestion(
+                    suggestion_text,
+                    rest.iter().map(|attr| (attr.span, String::new())).collect(),
+                    Applicability::MachineApplicable,
+                )
+                .emit();
+
+            return Err(());
         }
-        _ => cx.span_bug(trait_span, "method in `derive(Default)`"),
+    };
+    if !attr.is_word() {
```

Comment:
If `default` is introduced as a built-in attribute, then this checking will be performed automatically by `template!(Word)`.

Reply:
And if it's introduced as a helper attribute? As would ideally be the case.

Reply:
Then this logic will need to stay.